rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+
+        pub unsafe fn required_pages() -> u64 {
+            STATIC_MEM_IN_PAGES
+                + (super::total_allocated_blocks::get() as u64) * (PAGES_IN_BLOCK as u64)
+        }
+    }
+
+    /// Offsets into stable memory for statically-sized fields and tables.
+    pub mod offset {
+        pub const TOTAL_ALLOCATED_BLOCKS: u64 = 0;
+
+        pub const TOTAL_ALLOCATED_REGIONS: u64 = 2;
+
+        pub const BLOCK_REGION_TABLE: u64 = 4;
+
+        pub const REGION_TABLE: u64 = BLOCK_REGION_TABLE + super::size::BLOCK_REGION_TABLE;
+
+        /* One block for meta data, plus future use TBD. */
+        pub const BLOCK_ZERO: u64 = super::size::BLOCK_IN_BYTES;
+    }
+
+    pub mod total_allocated_blocks {
+        use super::offset;
+        use crate::ic0_stable::nicer::{read_u16, write_u16};
+
+        use crate::region::{REGION_SET_MEM_SIZE, REGION_TOTAL_ALLOCATED_BLOCKS};
+
+        pub fn get() -> u64 {
+            read_u16(offset::TOTAL_ALLOCATED_BLOCKS) as u64
+        }
+        pub fn set(n: u64) {
+            // Here we keep these copies of the total in sync.
+            //
+            // NB. The non-stable one is used when the stable one is
+            // unavailable (temp relocated by stable variable
+            // serialization/deserialization).
+            unsafe {
+                REGION_TOTAL_ALLOCATED_BLOCKS = n as u16;
+
+                // Invalidate stale number (No longer use the code-gen
+                // provided number that we need to use, temporarily,
+                // for destabilization after an upgrade).
+                REGION_SET_MEM_SIZE = None;
+            };
+            write_u16(offset::TOTAL_ALLOCATED_BLOCKS, n as u16)
+        }
+    }
+
+    pub mod total_allocated_regions {
+        use super::offset;
+        use crate::ic0_stable::nicer::{read_u16, write_u16};
+
+        pub fn get() -> u64 {
+            read_u16(offset::TOTAL_ALLOCATED_REGIONS) as u64
+        }
+        pub fn set(n: u64) {
+            write_u16(offset::TOTAL_ALLOCATED_REGIONS, n as u16)
+        }
+    }
+
+    pub mod block_region_table {
+        // invariant: all blocks whose IDs are below the total_allocated_blocks are valid.
+
+        use super::{offset, size};
+        use crate::ic0_stable::nicer::{read_u16, write_u16};
+        use crate::region::{BlockId, RegionId};
+
+        // Compute an offset in stable memory for a particular block ID (based zero).
+        fn index(b: &BlockId) -> u64 {
+            offset::BLOCK_REGION_TABLE + (b.0 as u64 * size::BLOCK_REGION_TABLE_ENTRY as u64)
+        }
+
+        /// Some(r, j) means that the block is in use by region r, at slot j.
+        /// None means that the block is available for (re-)allocation.
+        pub fn get(b: BlockId) -> Option<(RegionId, u16)> {
+            let raw = read_u16(index(&b));
+            let rid = RegionId::from_u16(raw);
+            rid.map(|r| (r, read_u16(index(&b) + 2)))
```

Comment:
Detail: Maybe the `2` in `get` and `set` could be replaced by `size_of::<u16>()`, or maybe an offset constant for both fields inside the table element.

Reply:
Done (I think)